High contrast theme

aaronshekey opened the pull request:
This PR explores adding high contrast modes in both light and dark modes. We'll likely need to get this in front of some users to see if it's even a net positive.

Interestingly, with a cursory search, it doesn't appear that there's a reliable media query to auto-select high contrast modes 😔

- [x] Syntax highlighting in dark mode
- [x] Notice border colors

aaronshekey committed:
High contrast theme toggle

aaronshekey committed:
Basic light high contrast theme

aaronshekey committed:
Add a basic dark high contrast theme

aaronshekey committed:
Better oranges

aaronshekey committed:
Better lighter blacks

aaronshekey committed:
Merge branch 'develop' into feature/high-contrast-theme

aaronshekey committed:
High contrast dark mode syntax highlighting

aaronshekey committed:
Notice overrides

aaronshekey committed:
Merge branch 'develop' into feature/high-contrast-theme

aaronshekey committed:
Remove comments

aaronshekey committed:
Syntax

aaronshekey committed:
Merge branch 'develop' into feature/high-contrast-theme

dancormier committed:
Merge branch 'develop' into feature/high-contrast-theme

dancormier commented:
With develop merged in, I think we have a decent path forward here. Long term, I'd love to make the color generation programmatic but that's a project for another time.

I dug around using [Firefox's _contrast loss_ mode](https://developer.mozilla.org/en-US/docs/Tools/Accessibility_inspector/Simulation) to pick out some contrast issues I'd like to resolve before merging:

- [ ] [Banner](https://deploy-preview-556--stacks.netlify.app/product/components/banners/) - Bottom border too light
    <details>
    <summary>Screenshot</summary>

    ![image](https://user-images.githubusercontent.com/647177/132724457-5c0095b1-5116-4505-a7d4-1c0c8dc1fb2b.png)

    </details>

- [x] [Button group](https://deploy-preview-556--stacks.netlify.app/product/components/banners/) - Selected hard to differentiate from unselected
    <details>
    <summary>Screenshot</summary>

    ![image](https://user-images.githubusercontent.com/647177/132725333-fdfb748d-87c4-43bd-9c99-0b74c8ad9f43.png)

    </details>

- [x] [Label > Status](https://deploy-preview-556--stacks.netlify.app/product/components/labels/#required) - Contrast lower in high contrast mode
    <details>
    <summary>Screenshot</summary>

    ![image](https://user-images.githubusercontent.com/647177/132725752-25f39eb0-9905-4c44-b87a-2285d0234ad7.png)

    </details>

- [x] [Navigation](https://deploy-preview-556--stacks.netlify.app/product/components/navigation/#default) - Selected state
    <details>
    <summary>Screenshot</summary>

    ![image](https://user-images.githubusercontent.com/647177/132726042-9975751f-007f-497e-85f5-5f5a260c593c.png)

    </details>

- [x] [Notices > Filled](https://deploy-preview-556--stacks.netlify.app/product/components/notices/#filled) - Contrast lower in high contrast mode
    <details>
    <summary>Screenshot</summary>

    ![image](https://user-images.githubusercontent.com/647177/132726236-e1766d66-9380-4d45-84a2-7fab98e5e099.png)

    </details>

- [x] [Notices > Toast](https://deploy-preview-556--stacks.netlify.app/product/components/notices/#toast) - Contrast lower in high contrast mode
    <details>
    <summary>Screenshot</summary>

    ![image](https://user-images.githubusercontent.com/647177/132726430-5be339b9-8e4f-4641-a5d4-2859a0d70996.png)

    </details>

- [x] [Pagination](https://deploy-preview-556--stacks.netlify.app/product/components/pagination/#example) - Selected state
    <details>
    <summary>Screenshot</summary>

    ![image](https://user-images.githubusercontent.com/647177/132726612-3cb33fb3-c59d-427d-a109-9dc8cc4f93da.png)

    </details>

- [x] [Progress bars > Fill alternates](https://deploy-preview-556--stacks.netlify.app/product/components/progress-bars/#fill-alternates) - Contrast lower in high contrast mode
    <details>
    <summary>Screenshot</summary>

    ![image](https://user-images.githubusercontent.com/647177/132727009-2e3dcaa2-a57a-41b2-a170-8f92d5d39db2.png)

    </details>

- [ ] [Progress bars > Segmented](https://deploy-preview-556--stacks.netlify.app/product/components/progress-bars/#segmented) - Contrast lower in high contrast mode
    <details>
    <summary>Screenshot</summary>

    ![image](https://user-images.githubusercontent.com/647177/132727040-50d03d7a-cc23-4ef2-acdb-aaff6b891e3a.png)

    </details>

- [x] [Sidebar widgets - Alternative colors](https://deploy-preview-556--stacks.netlify.app/product/components/sidebar-widgets/#alternative-colors) - Contrast lower in high contrast mode
    <details>
    <summary>Screenshot</summary>

    ![image](https://user-images.githubusercontent.com/647177/132727064-e0b77d27-a3db-4d3b-ae17-dc10408e22c5.png)

    </details>

- [x] [Tags](https://deploy-preview-556--stacks.netlify.app/product/components/tags/) - Contrast lower in high contrast mode
    <details>
    <summary>Screenshot</summary>

    ![image](https://user-images.githubusercontent.com/647177/132727113-b83a26a3-df1f-4780-a0c2-a3f88943ae18.png)

    </details>

- [x] Theme colors - Theme colors aren't modified in high contrast mode

dancormier committed:
OOPS! All theme popover buttons

dancormier committed:
Fix dark mode; shuffle some deck chairs

dancormier committed:
High Contrast Theme: Part II (#752)

* Standardize high contrast black; add bc variables

* Fix import notice high contrast background color

* Standardize high contrast colors

This standardizes high contrast mode colors roughly along perceived lightness (not just hsl lightness value). I took some liberties with gold and yellow, but nearly all colors are either > 85% or < 35% lightness.

* Tweak button, badge high contrast colors

* Update examples using deprecated `.*-black-1/2/3`

* dark mode play nice with high contrast

* Remove unneeded orange/blue hc values

* Use variable for scrollbar color

* Ensure dark hc mode shows in forced dark modes

* Fix hc mode typo

* Update dark hc color values

* Improve dark hc borders/box-shadow

* Add `highcontrast-mode` mixin

* Tweak button contrast

* Fix dark hc button number color

* fix hc s-input-icon colors

* Add border to s-status--label

* Add border to hc filled .s-notice

* Add --bc-lightest; standardize border colors

* Improve contrast on .s-staticwidget

* Improve tag contrast

* Improve table disabled contrast

* Improve toggle switches contrast

* Improve contrast for selected button in group

* Improve `.s-progress__privilege` bar contrast

* Make 300/400 hc division; juice colors

* Minor contrast tweaks

* Remove box shadow from toast example

* convert syntax highlight colors to hsl

* Add high contrast theme override

* Update theme default colors

* Tweak color docs contrast

* Tweak dark hc red, powder

* Tweak @tags-selected-background value

* Fix hc avatar colors

* Fix hc metal badges contrast

* Tweak hc dark mode blues

* Tweak hc metal colors

* Tweak yellow/orange to just barely meet AA

These colors are so tough to work with. For the love of god, just please don't use yellow.

* Use currentColor for hc button borders

This will help us meet WCAG Non-text contrast
https://www.w3.org/WAI/WCAG21/Understanding/non-text-contrast.html

* Better button borders

* Improve hc badge borders

* Improve hc badge border contrast

* Improve notice hc border

* Less of a cluster on hc btn border specificity

* Fix dark theme border color vars

* Update border classes b/w contrast

* Update navigation hc theme

* Get rid of drop shadows in hc mode

* Add border (as box-shadow) to avatar

* Juice input focus contrast

* Add hc breadcrumb divider color

* Improve uploader border contrast

* Fix the popover arrows

* Sharpen up the top arrows

Co-authored-by: Aaron Shekey <aaron@aaronshekey.com>

aaronshekey committed:
Refinements to syntax highlighting

aaronshekey committed:
Juice the contrast on blockquotes

aaronshekey committed:
Give dark mode highlight background some structure

aaronshekey committed:
Remove docs override

aaronshekey committed:
Merge branch 'develop' into feature/high-contrast-theme

aaronshekey committed:
Refine horizontal rules

aaronshekey commented:
This now has #752 in it, and can serve as the PR moving forward.

**TODO**
- [x] The homepage allows for setting a custom theme and high contrast

b-kelly commented on docs/_includes/layouts/home.html:
```diff
@@ -187,6 +187,25 @@
+                    <div class="s-popover--arrow"></div>
+                    <div class="grid fd-column gs12 gsy">
+                        <button class="grid--cell ta-left s-btn__unset fc-black-600 c-pointer js-darkmode-btn">Toggle dark mode</button>
+                        <button class="grid--cell ta-left s-btn__unset fc-black-600 c-pointer js-highcontrast-btn">Toggle high contrast</button>
```
Is this supposed to be here? Looks like an accidental partial copy of the popover above (has the same id, etc).

```suggestion
```

b-kelly commented on docs/_includes/layouts/page-nomenu.html:
```diff
@@ -0,0 +1,36 @@
+<!DOCTYPE html>
```
Is this still needed?

b-kelly committed:
fix forced light mode still using custom themes

b-kelly reviewed (commented):
Code looks good! Still doing some manual testing + size checking, so not approving juuuust yet.

b-kelly reviewed (commented):
Here's my size comparison. I ran:

```
> npm run build
> cd dist/css
> gzip -k stacks*.css
```
I then compared the sizes of each resulting `.gz` file as reported by Windows explorer.

**develop branch**
* stacks.css: 52.7 KB (53,981 bytes)
* stacks.min.css: 48.3 KB (49,481 bytes)

**high contrast**
* stacks.css: 56.8 KB (58,206 bytes)
* stacks.min.css: 51.6 KB (52,839 bytes)

This results in a size increase of:
* stacks.css: 4225 bytes, ~8%
* stacks.min.css: 3358 bytes, ~7%

This size increase is kinda heavy, but not exorbitantly so. I'm going to do some quick checks on the outputted bundle to see if I can't pull some tricks to get it a bit smaller.

**Edit**

I made some tweaks to trim back the output as much as I'm able. The bulk of the new bytes comes from the new color variables, and there's no way for me to cut that out without cutting out the entire feature.

New sizes are as follows:

* stacks.css: 55.3 KB (56,651 bytes)
* stacks.min.css: 50.6 KB (51,859 bytes)

This results in a size increase of:
* stacks.css: 2670 bytes, ~5%
* stacks.min.css: 2378 bytes, ~5%

b-kelly committed:
remove unnecessary output bloat

b-kelly committed:
output tweak

b-kelly committed:
add highcontrast dark focus rings

b-kelly committed:
share highcontrast themed colors

b-kelly reviewed (approved):
This looks awesome! Excellent work. This was a huge undertaking and it looks like you pretty much nailed it (to my non-designer, non-a11y expert eyes).

aaronshekey committed:
Remove merge conflict

aaronshekey committed:
Delete page-nomenu.html

aaronshekey committed:
A note on high contrast

aaronshekey committed:
Boost contrast of the stepped progress bar

aaronshekey committed:
Orange AAA

aaronshekey commented:
Ran Backstop and didn't see anything unexpected. There were some changes with respect to how colors are documented and examples in popovers, modals, and notices. Calling it! Merging.
